presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFlattenFunction.java

```diff
@@ -0,0 +1,91 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.scalar;
+
+import com.facebook.presto.metadata.FunctionInfo;
+import com.facebook.presto.metadata.FunctionRegistry;
+import com.facebook.presto.metadata.ParametricScalar;
+import com.facebook.presto.metadata.Signature;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeManager;
+import com.facebook.presto.spi.type.TypeSignature;
+import com.google.common.collect.ImmutableList;
+
+import java.lang.invoke.MethodHandle;
+import java.util.Map;
+
+import static com.facebook.presto.metadata.Signature.typeParameter;
+import static com.facebook.presto.type.TypeUtils.parameterizedTypeName;
+import static com.facebook.presto.util.Reflection.methodHandle;
+
+public class ArrayFlattenFunction
+        extends ParametricScalar
+{
+    public static final ArrayFlattenFunction ARRAY_FLATTEN_FUNCTION = new ArrayFlattenFunction();
+    private static final String FUNCTION_NAME = "array_flatten";
+    private static final Signature SIGNATURE = new Signature(FUNCTION_NAME, ImmutableList.of(typeParameter("E")), "E", ImmutableList.of("array<E>"), false, false);
+    private static final MethodHandle METHOD_HANDLE = methodHandle(ArrayFlattenFunction.class, FUNCTION_NAME, Type.class, Block.class);
+
+    @Override
+    public Signature getSignature()
+    {
+        return SIGNATURE;
+    }
+
+    @Override
+    public boolean isHidden()
+    {
+        return false;
+    }
+
+    @Override
+    public boolean isDeterministic()
+    {
+        return true;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "Flattens a given array";
+    }
+
+    @Override
+    public FunctionInfo specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        Type elementType = types.get("E");
+        MethodHandle methodHandle = METHOD_HANDLE.bindTo(elementType);
+        TypeSignature typeSignature = parameterizedTypeName("array", types.get("E").getTypeSignature());
+        Signature signature = new Signature(FUNCTION_NAME, elementType.getTypeSignature(), typeSignature);
+        return new FunctionInfo(signature, getDescription(), isHidden(), methodHandle, isDeterministic(), false, ImmutableList.of(false));
+    }
+
+    public static Block array_flatten(Type type, Block array)
+    {
+        boolean initialized = false;
+        Block resultArray = null;
+        for (int i = 0; i < array.getPositionCount(); ++i) {
+            Block block = (Block) type.getObject(array, i);
+            if (!initialized) {
+                resultArray = block;
+                initialized = true;
+            }
+            else {
+                resultArray = ArrayConcatUtils.concat(type.getTypeParameters().get(0), resultArray, block);
```

Comment:
The way that I would suggest implementing this would be to bind `type` to the `E` type in `specialize` then call `type.createBlockBuilder` to get a block builder, then use a double nested loop over all the blocks in `array` and use `type.appendTo` to add them to the `BlockBuilder`